state-chain/pallets/cf-pools/src/lib.rs

```diff
@@ -40,182 +40,205 @@ mod tests;
 
 impl_pallet_safe_mode!(PalletSafeMode; range_order_update_enabled, limit_order_update_enabled);
 
-/// For referring to either the stable or unstable asset of a particular pool
-enum Stability {
-	Stable,
-	Unstable,
-}
-
 // TODO Add custom serialize/deserialize and encode/decode implementations that preserve canonical
 // nature.
-/// Represents a pair of assets in a canonical ordering, so given two different assets they are
-/// always the same way around. In this case the unstable asset is `zero` and the stable is `one`.
+// TODO Move into pallet
 #[derive(Copy, Clone, Debug, Encode, Decode, TypeInfo, MaxEncodedLen, PartialEq, Eq)]
-pub struct CanonicalAssetPair {
-	assets: cf_amm::common::SideMap<Asset>,
+pub struct AssetPair {
+	assets: AssetsMap<Asset>,
 }
-impl CanonicalAssetPair {
-	pub fn new(base_asset: Asset, pair_asset: Asset) -> Option<Self> {
-		match (base_asset, pair_asset) {
-			(STABLE_ASSET, STABLE_ASSET) => None,
-			(STABLE_ASSET, unstable_asset) | (unstable_asset, STABLE_ASSET) => Some(Self {
-				assets: cf_amm::common::SideMap::<()>::default().map(|side, _| {
-					match Self::side_to_stability(side) {
-						Stability::Stable => STABLE_ASSET,
-						Stability::Unstable => unstable_asset,
-					}
-				}),
-			}),
-			_ => None,
-		}
+impl AssetPair {
+	pub fn new(base_asset: Asset, quote_asset: Asset) -> Option<Self> {
+		Some(AssetPair {
+			assets: match (base_asset, quote_asset) {
+				(STABLE_ASSET, STABLE_ASSET) => None,
+				(_unstable_asset, STABLE_ASSET) =>
+					Some(AssetsMap { base: base_asset, quote: quote_asset }),
+				_ => None,
+			}?,
+		})
 	}
 
-	pub fn new_with_result<T: Config>(
-		base_asset: Asset,
-		pair_asset: Asset,
-	) -> Result<Self, Error<T>> {
-		Self::new(base_asset, pair_asset).ok_or(Error::<T>::PoolDoesNotExist)
+	pub fn try_new<T: Config>(base_asset: Asset, quote_asset: Asset) -> Result<Self, Error<T>> {
+		Self::new(base_asset, quote_asset).ok_or(Error::<T>::PoolDoesNotExist)
 	}
 
-	fn side_to_asset(&self, side: Side) -> Asset {
-		self.assets[side]
+	pub fn from_swap(from: Asset, to: Asset) -> Option<(Self, Order)> {
+		#[allow(clippy::manual_map)]
+		if let Some(asset_pair) = Self::new(from, to) {
+			Some((asset_pair, Order::Sell))
+		} else if let Some(asset_pair) = Self::new(to, from) {
+			Some((asset_pair, Order::Buy))
+		} else {
+			None
+		}
 	}
 
-	/// !!! Must match side_to_stability !!!
-	fn stability_to_side(stability: Stability) -> Side {
-		match stability {
-			Stability::Stable => Side::One,
-			Stability::Unstable => Side::Zero,
-		}
+	pub fn assets(&self) -> AssetsMap<Asset> {
+		self.assets
 	}
+}
+
+#[derive(
+	Copy,
+	Clone,
+	Debug,
+	Encode,
+	Decode,
+	TypeInfo,
+	MaxEncodedLen,
+	PartialEq,
+	Eq,
+	Deserialize,
+	Serialize,
+)]
+pub enum Assets {
```

Comment:
This is also kind of a confusing name...

Reply:
We have have OrderSide (buy/sell), PoolSide (base/quote), AmmSide (one,zero) to disambiguate all of this a bit more clearly?

Reply:
I did give some better names 😅 
But ok to leave it if it's not exposed. What about order? This is one is in the Api.

Reply:
I know we want to get this out but once we settle on external-facing terminology it will be harder to change it.

Reply:
Not sure what you're suggesting here...

Reply:
<img width="790" alt="image" src="https://github.com/chainflip-io/chainflip-backend/assets/42404303/f1f4a38c-b83b-452e-bbdb-8497ae75e862">

GitHUb decided to not show your comments